### PR TITLE
Fix TUI transcript finalization

### PR DESCRIPTION
**Pull request by lsm**
## Summary
- Render final `message_end.text` into TUI transcript when streaming deltas are absent or incomplete
- Preserve progressive `text_delta` rendering while replacing partial text with exact final assistant content
- Add state tests covering final-only responses and duplicate-free streamed responses

## Test plan
- [x] `zig build -Doptimize=Debug --build-file zig/build.zig test-unit-tui`
- [x] `zig build -Doptimize=Debug --build-file zig/build.zig test`
- [x] `zig build -Doptimize=Debug --build-file zig/build.zig run-tui` launched TUI and rendered initial screen; command then terminated with SIGTERM to exit noninteractive run

### CODE REVIEW

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: ✅ APPROVE

**Single-file change** in `zig/src/tui/state.zig` — wires `message_end` into the transcript so final-only provider responses render in the TUI.

**Logic of `finishTranscriptEntry(kind, text)`:**
1. Empty text → early return (preserves streamed deltas for tool-call-only responses)
2. Last entry matches kind + identical text → no-op
3. Last entry matches kind + different text → clear + replace (final text is authoritative)
4. No matching trailing entry → append new entry

**Tests added (4 scenarios):**
- Final-only (no prior deltas, message_end creates entry)
- Streamed + identical final (hello deltas + hello final → 1 entry, no dup)
- Streamed + different final (hello delta → hello world final → replacement)
- User message echo (appendUserMessage + message_end same text → no dup)

All `MessageRole` variants handled exhaustively (assistant, user, tool_result).

No P0–P3 findings. Clean, minimal, well-tested.

File: zig/src/tui/state.zig [hunk 1]
```diff
@@ -432,6 +436,18 @@ pub const AppState = struct {
         try self.transcript.items[self.transcript.items.len - 1].text.appendSlice(self.allocator, delta);
     }
 
+    fn finishTranscriptEntry(self: *AppState, kind: TranscriptKind, text: []const u8) !void {
+        if (text.len == 0) return;
+        if (self.transcript.items.len > 0 and self.transcript.items[self.transcript.items.len - 1].kind == kind) {
+            const entry = &self.transcript.items[self.transcript.items.len - 1];
+            if (std.mem.eql(u8, entry.text.items, text)) return;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/state.zig]: Fixed by tracking active assistant transcript entry index from message_start/text_delta. Assistant message_end now updates that entry even when tool_call_delta created trailing tool row, so no extra assistant line or stale partial remains.

File: zig/src/tui/state.zig [hunk 2]
```diff
@@ -319,7 +319,11 @@ pub const AppState = struct {
             .text_delta => |payload| try self.appendDelta(.assistant, payload.delta.slice()),
             .thinking_delta => |payload| try self.appendDelta(.thinking, payload.delta.slice()),
             .tool_call_delta => |payload| try self.appendDelta(.tool, payload.delta.slice()),
-            .message_end => {},
+            .message_end => |payload| switch (payload.role) {
+                .assistant => try self.finishTranscriptEntry(.assistant, payload.text.slice()),
+                .user => try self.finishTranscriptEntry(.user, payload.text.slice()),
+                .tool_result => try self.finishTranscriptEntry(.tool, payload.text.slice()),
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/state.zig]: Fixed by tracking active tool-result transcript entry separately and only finalizing that indexed entry. message_start for tool_result now creates its own row, so tool execution rows are not clobbered by batched result finalization.

File: zig/src/tui/state.zig [hunk 3]
```diff
@@ -432,6 +436,18 @@ pub const AppState = struct {
...
     }
 
+    fn finishTranscriptEntry(self: *AppState, kind: TranscriptKind, text: []const u8) !void {
+        if (text.len == 0) return;
+        if (self.transcript.items.len > 0 and self.transcript.items[self.transcript.items.len - 1].kind == kind) {
+            const entry = &self.transcript.items[self.transcript.items.len - 1];
+            if (std.mem.eql(u8, entry.text.items, text)) return;
+            entry.text.clearRetainingCapacity();
+            try entry.text.appendSlice(self.allocator, text);
+            return;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/state.zig]: Fixed by using active entry indices instead of last-kind replacement. Inline message_end without a preceding message_start now appends a new transcript row instead of rewriting the previous assistant response.

**Review by lsm — COMMENTED** [review]
## 🤖 Review by Claude Sonnet 4 (Anthropic)

> **Model:** claude-sonnet-4-20250514 | **Client:** NeoKai | **Provider:** Anthropic

### Verdict: ✅ APPROVE (Round 2)

All 3 prior P2 findings resolved:

| Finding | Fix | Test |
|---------|-----|------|
| **Assistant entry lost after tool_call_delta** | `active_assistant_entry` index tracked from `message_start` / `text_delta`; `message_end` targets that index regardless of trailing entries | `"message_end updates active assistant before trailing tool"` |
| **Tool result clobbers tool execution rows** | `active_tool_result_entry` tracked separately; `message_start.tool_result` creates its own entry via `appendEmptyTranscript` | `"tool_result message_end updates active tool entry only"` |
| **Message_end-only replaces prior assistant** | `finishTranscriptEntry` uses active entry index; when null (no `message_start`), falls through to append-only path | `"message_end-only assistant appends after prior assistant"` |

**Design:** Index-based tracking (`?usize`) is the right abstraction — entries are append-only so indices remain stable. Bounds + kind checks in `finishTranscriptEntry` provide defense against stale indices (e.g., after `clearTranscript`).

Zero P0–P3 findings. Ship it.

File: zig/src/tui/state.zig [hunk 4]
```diff
@@ -422,14 +428,54 @@ pub const AppState = struct {
...
+                if (index < self.transcript.items.len and self.transcript.items[index].kind == kind) {
+                    try self.replaceEntryText(index, text);
+                    entry.* = null;
+                    return;
+                }
+            }
+            entry.* = null;
+        }
+
+        if (self.transcript.items.len > 0 and self.transcript.items[self.transcript.items.len - 1].kind == kind and std.mem.eql(u8, self.transcript.items[self.transcript.items.len - 1].text.items, text)) return;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/state.zig:470]: Fixed by removing assistant inline fallback dedupe. message_end-only assistant events now always append when no active assistant entry exists, preserving identical consecutive completions like repeated 'Done'. Added regression test.

File: zig/src/tui/state.zig [hunk 5]
```diff
@@ -422,14 +428,54 @@ pub const AppState = struct {
...
+        if (text.len == 0) {
+            if (active_entry) |entry| entry.* = null;
+            return;
+        }
+
+        if (active_entry) |entry| {
+            if (entry.*) |index| {
+                if (index < self.transcript.items.len and self.transcript.items[index].kind == kind) {
+                    try self.replaceEntryText(index, text);
+                    entry.* = null;
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/state.zig:463]: Fixed by clearing active transcript entry indices on turn_start, turn_end, agent_end, and error events. Stale active assistant indices cannot survive failed/interrupted turns and rewrite later inline completions. Added regression test.

File: zig/src/tui/state.zig [hunk 6]
```diff
@@ -312,14 +314,18 @@ pub const AppState = struct {
                 self.status.turn_count += 1;
             },
             .message_start => |payload| switch (payload.role) {
-                .assistant => try self.ensureTrailingEntry(.assistant),
-                .user => try self.ensureTrailingEntry(.user),
-                .tool_result => try self.ensureTrailingEntry(.tool),
+                .assistant => self.active_assistant_entry = try self.ensureTrailingEntry(.assistant),
```
[the remark this answers is not in the document]
Comment (lsm) [reply zig/src/tui/state.zig:317]: Fixed by making assistant message_start open a fresh assistant transcript row instead of reusing a trailing assistant row. message_end without deltas now finalizes that new row and preserves previous responses. Added regression test.
